### ISSUE TITLE
Build mechanistic distillation harness for the 2B→9B CED conversion under a 50 A100-hour budget

### ISSUE THREAD

**Issue by iamdarshg** [iamdarshg/KestrelVL-7B#9]
Parent epic: #2
Depends on: #3, #4
Integrates with: #5, #6, #7, #8

## Goal

Build the training/distillation harness that adapts the heterogeneous Qwen3.5-2B encoder + Qwen3.5-9B decoder architecture **without scratch pretraining** and enforces a hard first-run budget of approximately **50 A100-80GB GPU-hours**.

The modified fork itself is the student. Frozen pretrained behavior is the teacher/reference.

## Training principle

Do not rely only on next-token KL. Support mechanistic matching of the specific states introduced by architecture surgery.

Reference objective:

```text
L = λ_ce * CE
  + λ_kl * logit_KL
  + λ_h  * hidden_state_match
  + λ_a  * attention/output_match
  + λ_kv * memory/KV_match
  + λ_i  * index/retrieval_match
```

All terms and weights must be configurable and individually logged.

## Progressive training schedule

Initial budget target:

```text
0–5 h    baselines + bridge-only alignment
5–15 h   bridge + shared global KV
15–25 h  CSA2/indexer distillation
25–36 h  CED LoRA/selective decoder recovery
36–43 h  selective unfreeze of the smallest necessary inherited parameter set
43–47 h  medium/longer-context adaptation
47–50 h  final benchmark/regression sweep
```

The harness must support stopping early and reallocating unused budget when a stage converges or fails.

## Freeze/unfreeze policy

Start with:

- encoder frozen;
- decoder frozen;
- bridge/indexer/new gates trainable;
- Engram disabled unless explicitly running its ablation.

Then progressively allow:

1. bridge/global-KV/indexer parameters;
2. LoRA or adapters on selected decoder attention/residual projections;
3. upper encoder layers if required;
4. narrowly selected inherited full-rank parameters only if evidence shows the cheaper stages cannot recover quality.

Do not globally unfreeze the whole model by default.

## Teacher/reference modes

Support at least:

- standalone Qwen3.5-9B teacher on the original full prompt;
- CED student consuming Qwen3.5-2B encoder memory;
- optional current-Kestrel teacher/reference where useful for architecture-retention comparisons.

Teacher activations should be collected selectively; do not retain every layer/token activation if not required.

## Budget enforcement

Track and persist:

- wall-clock training time;
- actual GPU-seconds/GPU-hours;
- tokens processed;
- teacher vs student compute;
- tokens/s;
- peak VRAM;
- restart/resume overhead;
- effective cost of each stage.

Add a hard configurable budget such as:

```yaml
compute_budget:
  max_gpu_hours: 50
  gpu_class: a100_80gb
  hard_stop: true
```

Resume must preserve consumed-budget accounting.

## Data strategy

Prioritize a compact mixture that probes preserved capability rather than attempting general pretraining again:

- general text/language;
- code;
- reasoning;
- tool/schema-like exact structure;
- retrieval/exact-copy examples;
- multimodal examples only after the text CED path is stable;
- medium/long-context examples only late in the run.

## Required evaluation during training

At every stage record:

- logit KL vs source decoder;
- hidden-state reconstruction at selected decoder layers;
- validation loss/perplexity;
- short-context generation quality;
- exact-copy/retrieval accuracy;
- code/reasoning regression subset;
- prefill throughput;
- decode throughput;
- peak memory.

## Acceptance criteria

- [ ] Hard 50 A100-hour budget accounting is implemented and resume-safe.
- [ ] Bridge-only and selective-unfreeze stages can be run independently.
- [ ] Teacher/student hidden/logit/memory losses are available without full-activation retention.
- [ ] The system can recover useful decoder behavior without scratch pretraining.
- [ ] Every stage produces a checkpoint plus evidence report.
- [ ] The final report identifies exactly where the 50 GPU-hours were spent.
- [ ] If recovery is insufficient, the blocker is quantified rather than silently extending the compute budget.

**Comment by iamdarshg** [comment]
## Batch 1+2 CPU evidence — measurement skeleton PASS (no training runs, as scoped)

**Commits:** \8ac6f9c\ (\src/training/ced_distill.py\, \configs/model/qwen35_ced.yaml\) + KL clamp \15ef3ae\. **Tests:** \	ests/test_ced_distill.py\ — 20/20 pass on CPU. Full suite 174 green.

**Proof (instrumentation, all CPU/tiny tensors):**
- Reference objective implemented per-term: \compute_losses\ returns {ce, kl, hidden, attn, kv, index, total} + full \log_dict\ with per-term weights and \ *_active\ flags; zero when teacher==student, KL>0 on ramp perturbation (tested; uniform-shift-invariance correctly yields KL≈0).
- Selected-layers only: unselected layers ignored (tested); never full retention by default (selected [4,8,12,16,20]).
- Recovery gate: \measure_recovery\ quantifies {max_abs_diff, mean_abs_diff, kl, within_tol} — identical→True, perturbed→False (tested). KL clamped at 0.0 (was −3e−08 fp noise).
- Memory stats: norms/mean/std/rank-proxy + deterministic fingerprint (tested). Param accounting sums trainable/frozen/by-prefix (tested).
- Freeze policy: bridge_only / bridge_global_kv (enc+dec frozen) → lora_recovery → selective_unfreeze; unknown stage raises (tested).
- Budget: \BudgetTracker\ (50h, a100_80gb, hard_stop) accumulates per-stage, raises on overrun, save/load preserves consumed (tested). No 50h run launched — stages unrunnable until the GPU smoke test passes.
- Config: \qwen35_ced.yaml\ round-trips into DistillLossConfig/FreezePolicy/BudgetTracker (tested); revisions now pinned (2B b1485b2f, 9B c2022362).

**True-cost note:** .00 spent to date (ledger \
eports/gcp-cost-ledger.jsonl\). Next spend is the L4 spot smoke test, pending approval — its actual billed cost will be reconciled into this issue + ledger after the run.

**Comment by iamdarshg** [comment]
## True-cost report: L4 smoke campaign (7 attempts, 2026-09-13)

**Estimated total: ~.92** (per-second on-demand billing @ \.7068/h + PD; labeled \estimated\ — no billing export configured, so not billing-confirmed). Per-attempt ledger in \
eports/gcp-cost-ledger.jsonl\: \.01 + \.01 + \.06 + \.16 + \.26 + \.12 + **\.30 (PASS)**. Honest note: cumulative exceeds the original \.82 single-run framing — 4 of 7 attempts were fast-fail diagnostics that each bought a root cause (interpreter/packaging, spot preemption, vocab proof, torchaudio ABI).

Cheapest-adequate thesis held: L4 on-demand total < one A100 hour. No VM remains running (all smoke instances deleted; verified via \gcloud compute instances list\).
